core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -306,21 +321,68 @@ public synchronized void close() {
     mClosed = true;
   }
 
+  /**
+   *  {@link AbstractClient} works with Grpc Servers.
+   *  Child classes should only override this method to query the address
+   *  of the grpc server they talk to. The conversion from {@link GrpcServerAddress}
+   *  to more generic {@link SocketAddress} required by {@link Client#getRemoteSockAddress()}
+   *  is handled by this class.
+   *
+   * @return the {@link GrpcServerAddress} of the remote server
+   * @throws UnavailableException if address can't be determined
+   */
+  protected abstract GrpcServerAddress queryGrpcServerAddress() throws UnavailableException;
```

Comment:
Seems you rely on the caller to synchronize so this call does not have to be `synchronized`. But if that's the case I actually think you can make this one `synchronized` too, for better readability. So the next dummy like me need not think why only this one is not synchronized. WDYT?

Reply:
Unfortunately abstract method can't be declared synchronized at the same time. I'd think the synchronization is handled by the framework methods in `AbstractClient`, so that whether the components abstract methods are synchronized themselves are unimportant.

Reply:
Maybe add an annotation `@GuardedBy("this")`?

Reply:
Actually I think the semantics of `AbstractClient` doesn't require the abstract method to be synchronized anyway.  It's almost always used in synchronized wrappers, i.e., `connect` and `retryRPC`, just like `beforeConnect` and `afterConnect`. If a child class decides to bypass the parent and call these methods directly, then it is its own responsibility to deal with concurrency issues thus arised.

Reply:
The current override of `queryGrpcServerAddress` in `AbstractMasterClient` is synchronized, but it could well be not. Maybe I'm being lazy to rely on the re-entrance lock of java's synchronized keyword.